File: common/src/main/java/org/seasar/doma/it/IntegrationTestEnvironment.java
package org.seasar.doma.it;

import static org.junit.jupiter.api.extension.ConditionEvaluationResult.disabled;
import static org.junit.jupiter.api.extension.ConditionEvaluationResult.enabled;
import static org.junit.platform.commons.util.AnnotationUtils.findAnnotation;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.junit.jupiter.api.extension.AfterAllCallback;
import org.junit.jupiter.api.extension.AfterTestExecutionCallback;
import org.junit.jupiter.api.extension.BeforeAllCallback;
import org.junit.jupiter.api.extension.BeforeTestExecutionCallback;
import org.junit.jupiter.api.extension.ConditionEvaluationResult;
import org.junit.jupiter.api.extension.ExecutionCondition;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.junit.jupiter.api.extension.ParameterContext;
import org.junit.jupiter.api.extension.ParameterResolutionException;
import org.junit.jupiter.api.extension.ParameterResolver;
import org.seasar.doma.it.dao.ScriptDao;
import org.seasar.doma.it.dao.ScriptDaoImpl;
import org.seasar.doma.jdbc.Config;
import org.seasar.doma.jdbc.dialect.Db2Dialect;
import org.seasar.doma.jdbc.dialect.Dialect;
import org.seasar.doma.jdbc.dialect.H2Dialect;
import org.seasar.doma.jdbc.dialect.HsqldbDialect;
import org.seasar.doma.jdbc.dialect.MssqlDialect;
import org.seasar.doma.jdbc.dialect.MysqlDialect;
import org.seasar.doma.jdbc.dialect.OracleDialect;
import org.seasar.doma.jdbc.dialect.PostgresDialect;
import org.seasar.doma.jdbc.dialect.SqliteDialect;

public class IntegrationTestEnvironment
    implements BeforeAllCallback,
        BeforeTestExecutionCallback,
        AfterTestExecutionCallback,
        AfterAllCallback,
        ParameterResolver,
        ExecutionCondition {

  private static final Logger logger = Logger.getLogger(IntegrationTestEnvironment.class.getName());

  private static final Pattern jdbcUrlPattern = Pattern.compile("^jdbc:([^:]*):.*");

  private static transient boolean imported;

  private final Dbms dbms;

  private final AppConfig config;

  private final ScriptDao scriptDao;

  public IntegrationTestEnvironment() {
    String url = getProperty("url", "jdbc:h2:mem:doma_it;DB_CLOSE_DELAY=-1");
    logger.log(Level.INFO, "url=" + url);
    String user = getProperty("user", "sa");
    logger.log(Level.INFO, "user=" + user);
    String password = getProperty("password", "");
    dbms = determineDbms(url);
    Dialect dialect = createDialect(dbms);
    config = new AppConfig(dialect, dbms, url, user, password);
    scriptDao = new ScriptDaoImpl(config);
  }

  private static String getProperty(String key, String defaultValue) {
    String value = System.getProperty(key);
    if (value != null && value.length() > 0) {
      return value;
    }
    return defaultValue;
  }

  private static Dbms determineDbms(String url) {
    Matcher matcher = jdbcUrlPattern.matcher(url);
    if (matcher.matches()) {
      return Dbms.valueOf(matcher.group(1).toUpperCase());
    }
    throw new IllegalArgumentException("url: " + url);
  }

  private static Dialect createDialect(Dbms dbms) {
    switch (dbms) {
      case H2:
        return new H2Dialect();
      case HSQLDB:
        return new HsqldbDialect();
      case SQLITE:
        return new SqliteDialect();
      case MYSQL:
        return new MysqlDialect();
      case POSTGRESQL:
        return new PostgresDialect();
      case SQLSERVER:
        return new MssqlDialect();
      case ORACLE:
        return new OracleDialect();
      case DB2:
        return new Db2Dialect();
    }
    throw new IllegalArgumentException("unreachable: " + dbms);
  }

  @Override
  public void beforeAll(ExtensionContext context) {
    if (imported) {
      return;
    }
    config.getTransactionManager().required(scriptDao::create);
    imported = true;
  }

  @Override
  public void beforeTestExecution(ExtensionContext context) {
    config.getLocalTransaction().begin();
  }

  @Override
  public void afterTestExecution(ExtensionContext context) {
    config.getLocalTransaction().rollback();
  }

  @Override
  public void afterAll(ExtensionContext context) {}

  @Override
  public boolean supportsParameter(
      ParameterContext parameterContext, ExtensionContext extensionContext)
      throws ParameterResolutionException {
    return parameterContext.getParameter().getType() == Config.class;
  }

  @Override
  public Object resolveParameter(
      ParameterContext parameterContext, ExtensionContext extensionContext)
      throws ParameterResolutionException {
    return config;
  }

  @Override
  public ConditionEvaluationResult evaluateExecutionCondition(ExtensionContext context) {
    Optional<Run> optional = findAnnotation(context.getElement(), Run.class);
    if (!optional.isPresent()) {
      return enabled("@Run is not present");
    }
    Run run = optional.get();
    if (isRunnable(run, dbms)) {
      return enabled("runnable: " + dbms);
    }
    return disabled("not runnable");
  }

  private boolean isRunnable(Run run, Dbms dbms) {
    List<Dbms> onlyIf = Arrays.asList(run.onlyIf());
    List<Dbms> unless = Arrays.asList(run.unless());

    if (onlyIf.size() > 0) {
      return onlyIf.contains(dbms);
    } else {
      if (unless.size() > 0) {
        return !unless.contains(dbms);
      } else {
        return true;
      }
    }
  }
}

File: kotlin/src/test/kotlin/org/seasar/doma/it/kotlin/IntegrationTestEnvironment.kt
package org.seasar.doma.it.kotlin

import org.junit.jupiter.api.extension.AfterAllCallback
import org.junit.jupiter.api.extension.AfterTestExecutionCallback
import org.junit.jupiter.api.extension.BeforeAllCallback
import org.junit.jupiter.api.extension.BeforeTestExecutionCallback
import org.junit.jupiter.api.extension.ConditionEvaluationResult
import org.junit.jupiter.api.extension.ExecutionCondition
import org.junit.jupiter.api.extension.ExtensionContext
import org.junit.jupiter.api.extension.ParameterContext
import org.junit.jupiter.api.extension.ParameterResolutionException
import org.junit.jupiter.api.extension.ParameterResolver
import org.junit.platform.commons.util.AnnotationUtils
import org.seasar.doma.it.kotlin.dao.ScriptDao
import org.seasar.doma.it.kotlin.dao.ScriptDaoImpl
import org.seasar.doma.jdbc.Config
import org.seasar.doma.jdbc.dialect.Db2Dialect
import org.seasar.doma.jdbc.dialect.Dialect
import org.seasar.doma.jdbc.dialect.H2Dialect
import org.seasar.doma.jdbc.dialect.HsqldbDialect
import org.seasar.doma.jdbc.dialect.MssqlDialect
import org.seasar.doma.jdbc.dialect.MysqlDialect
import org.seasar.doma.jdbc.dialect.OracleDialect
import org.seasar.doma.jdbc.dialect.PostgresDialect
import org.seasar.doma.jdbc.dialect.SqliteDialect
import java.util.Optional
import java.util.logging.Level
import java.util.logging.Logger
import java.util.regex.Pattern

class IntegrationTestEnvironment : BeforeAllCallback, BeforeTestExecutionCallback, AfterTestExecutionCallback, AfterAllCallback, ParameterResolver, ExecutionCondition {
    private val dbms: Dbms
    private val config: AppConfig
    private val scriptDao: ScriptDao

    init {
        val url = getProperty("url", "jdbc:h2:mem:doma_it;DB_CLOSE_DELAY=-1")
        logger.log(Level.INFO, "url=$url")
        val user = getProperty("user", "sa")
        logger.log(Level.INFO, "user=$user")
        val password = getProperty("password", "")
        dbms = determineDbms(url)
        val dialect = createDialect(dbms)
        config = AppConfig(dialect, dbms, url, user, password)
        scriptDao = ScriptDaoImpl(config)
    }

    @Throws(Exception::class)
    override fun beforeAll(context: ExtensionContext) {
        if (imported) {
            return
        }
        config
            .getTransactionManager()
            .required(
                Runnable { scriptDao.create() }
            )
        imported = true
    }

    @Throws(Exception::class)
    override fun beforeTestExecution(context: ExtensionContext) {
        config.localTransaction.begin()
    }

    @Throws(Exception::class)
    override fun afterTestExecution(context: ExtensionContext) {
        config.localTransaction.rollback()
    }

    @Throws(Exception::class)
    override fun afterAll(context: ExtensionContext) {
    }

    @Throws(ParameterResolutionException::class)
    override fun supportsParameter(
        parameterContext: ParameterContext,
        extensionContext: ExtensionContext
    ): Boolean {
        return parameterContext.parameter.type == Config::class.java
    }

    @Throws(ParameterResolutionException::class)
    override fun resolveParameter(
        parameterContext: ParameterContext,
        extensionContext: ExtensionContext
    ): Any {
        return config
    }

    override fun evaluateExecutionCondition(context: ExtensionContext): ConditionEvaluationResult {
        val optional: Optional<Run> = AnnotationUtils.findAnnotation<Run>(context.element, Run::class.java)
        if (!optional.isPresent()) {
            return ConditionEvaluationResult.enabled("@Run is not present")
        }
        val run: Run = optional.get()
        return if (isRunnable(run, dbms)) {
            ConditionEvaluationResult.enabled("runnable: $dbms")
        } else ConditionEvaluationResult.disabled("not runnable")
    }

    private fun isRunnable(run: Run, dbms: Dbms): Boolean {
        val onlyIf: List<Dbms> = listOf(*run.onlyIf)
        val unless: List<Dbms> = listOf(*run.unless)
        return if (onlyIf.size > 0) {
            onlyIf.contains(dbms)
        } else {
            if (unless.size > 0) {
                !unless.contains(dbms)
            } else {
                true
            }
        }
    }

    companion object {
        private val logger = Logger.getLogger(IntegrationTestEnvironment::class.java.name)
        private val jdbcUrlPattern = Pattern.compile("^jdbc:([^:]*):.*")

        @Transient
        private var imported = false
        private fun getProperty(key: String, defaultValue: String): String {
            val value = System.getProperty(key)
            return if (value != null && value.length > 0) {
                value
            } else defaultValue
        }

        private fun determineDbms(url: String): Dbms {
            val matcher = jdbcUrlPattern.matcher(url)
            if (matcher.matches()) {
                return Dbms.valueOf(matcher.group(1).toUpperCase())
            }
            throw IllegalArgumentException("url: $url")
        }

        private fun createDialect(dbms: Dbms): Dialect {
            return when (dbms) {
                Dbms.H2 -> H2Dialect()
                Dbms.HSQLDB -> HsqldbDialect()
                Dbms.SQLITE -> SqliteDialect()
                Dbms.MYSQL -> MysqlDialect()
                Dbms.POSTGRESQL -> PostgresDialect()
                Dbms.SQLSERVER -> MssqlDialect()
                Dbms.ORACLE -> OracleDialect()
                Dbms.DB2 -> Db2Dialect()
            }
        }
    }
}
